### PR TITLE
[WIP] StudioResource pivot table

### PR DESCRIPTION
**Pull request by ProWeebDev**
This pivot class will allow a studio to have multiple external resources attached to it.

### CODE REVIEW

File: app/Console/Commands/Wiki/DatabaseDumpCommand.php [hunk 1]
```diff
@@ -85,6 +86,7 @@ class DatabaseDumpCommand extends Command
         Song::TABLE,
         Studio::TABLE,
         Video::TABLE,
+        StudioResource::TABLE,
```
Comment (paranarimasu) [inline app/Console/Commands/Wiki/DatabaseDumpCommand.php:89]: Can we keep this in alphabetical order?

Comment (ProWeebDev) [reply app/Console/Commands/Wiki/DatabaseDumpCommand.php:88]: Fixed

**Review by paranarimasu — CHANGES_REQUESTED** [review]
Initial review completed

File: app/Models/Wiki/Studio.php [hunk 2]
```diff
@@ -22,6 +23,7 @@
  * @property string $name
  * @property string $slug
  * @property int $studio_id
+ * @property BasePivot $pivot
```
Comment (paranarimasu) [inline app/Models/Wiki/Studio.php:26]: Undefined Class BasePivot. It needs to be imported.

File: app/Pivots/StudioResource.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+
+/**
+ * Class StudioResource.
+ *
+ * @property Studio $studio
+ * @property string $as
+ * @property ExternalResource $resource
+ *
+ * @method static StudioResourceFactory factory(...$parameters)
```
Comment (paranarimasu) [inline app/Pivots/StudioResource.php:21]: Undefined class StudioResourceFactory. It needs to be imported.

File: app/Nova/Resources/Wiki/Studio.php [hunk 4]
```diff
@@ -129,6 +130,22 @@ public function fields(Request $request): array
                     ];
                 }),
 
+            BelongsToMany::make(__('nova.external_resources', 'Resources', ExternalResource::class))
+                ->searchable()
+                ->fields(function () {
+                    return [
+                        Text::make(__('nova.as'), StudioResource::ATTRIBUTE_AS)
+                            ->rules(['nullable', 'max:192'], )
```
Comment (paranarimasu) [inline app/Nova/Resources/Wiki/Studio.php:138]: Code smell: useless trailing comma.

File: tests/Feature/Jobs/Pivot/StudioResourceTest.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        Bus::fake(SendDiscordNotificationJob::class);
+
+        $studio->resources()->attach($resource);
+
+        Bus::assertDispatched(SendDiscordNotificationJob::class);
+    }
+
+    /**
+     * When a Studio is detached from a Resource or vice versa, a SendDiscordNotification job shall be dispatched.
+     */
```
Comment (paranarimasu) [inline tests/Feature/Jobs/Pivot/StudioResourceTest.php:40]: Static Analysis: Add return typehint "@return void".

File: tests/Feature/Jobs/Pivot/StudioResourceTest.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        Bus::fake(SendDiscordNotificationJob::class);
+
+        $studio->resources()->detach($resource);
+
+        Bus::assertDispatched(SendDiscordNotificationJob::class);
+    }
+
+    /**
+     * When a Studio Resource pivot is updated, a SendDiscordNotification job will be dispatched.
+     */
```
Comment (paranarimasu) [inline tests/Feature/Jobs/Pivot/StudioResourceTest.php:58]: Static Analysis: Add return typehint "@return void".

File: app/Nova/Resources/Wiki/Studio.php [hunk 7]
```diff
@@ -129,6 +130,22 @@ public function fields(Request $request): array
                     ];
                 }),
 
+            BelongsToMany::make(__('nova.external_resources', 'Resources', ExternalResource::class))
```
Comment (paranarimasu) [inline app/Nova/Resources/Wiki/Studio.php:133]: Erroneous closing parenthesis at the end of the line. There should be a closing parenthesis after 'nova.external_resources' as this is the translation key for the __() translation helper function.

**Review by paranarimasu — CHANGES_REQUESTED** [review]
Second review. I followed through with code inspection, static analysis and test case runs.

File: app/Nova/Resources/Wiki/ExternalResource.php [hunk 8]
```diff
@@ -182,6 +183,24 @@ public function fields(Request $request): array
                     ];
                 }),
 
+            BelongsToMany::make(__('nova.studio'), 'Studio', Studio::class)
```
Comment (paranarimasu) [inline app/Nova/Resources/Wiki/ExternalResource.php:186]: Acceptance test finding: 'Studio' needs to be 'Studios'. This throws an exception when viewing an external resource in nova.

File: app/Http/Api/Schema/Wiki/StudioSchema.php [hunk 9]
```diff
@@ -37,6 +37,7 @@ public function allowedIncludes(): array
     {
         return [
             AllowedInclude::make(AnimeSchema::class, Studio::RELATION_ANIME),
+            AllowedInclude::make(ExternalResourceSchema::class, Studio::RELATION_RESOURCES),
         ];
     }
```
Comment (paranarimasu) [inline app/Http/Api/Schema/Wiki/StudioSchema.php:43]: We need to add the StudioResource::ATTRIBUTE_AS field to the fields function. As it is right now, attempting to select this field for sparse fieldsets produces a validation error. 

Please use the same approach as AnimeSchema does to include the pivot field.

**Review by paranarimasu — CHANGES_REQUESTED** [review]
Acceptance testing findings. This should be the final review.

**Review by paranarimasu — APPROVED** [review]
Code analysis passed. Static Analysis passed. Tests passed. Acceptance passed.

PR approved.

**Pull request merged**